sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -2867,6 +2867,19 @@ pi_result cuda_piEnqueueKernelLaunch(
       retImplEv->start();
     }
 
+    // Set local mem max size if env var is present
+    {
+      char *local_mem_sz_ptr;
+      if ((local_mem_sz_ptr = std::getenv("SYCL_PI_CUDA_MAX_LOCAL_MEM_SZ")) !=
+          nullptr) {
+        int val = std::atoi(local_mem_sz_ptr);
+        if (val) {
```

Comment:
This check means that an invalid value (or 0) specified in `SYCL_PI_CUDA_MAX_LOCAL_MEM_SZ` will be ignored silently. It would be better to return some error code here to allow the caller of `cuda_piEnqueueKernelLaunch` (or `piEnqueueKernelLaunch` by proxy) to throw an appropriate error. If that is the solution we go with, please also mention this in the documentation of the environment variable.

Likewise, it would be great to check if the value exceeds the allowed max shared memory, i.e. the [`CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK_OPTIN`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gge12b8a782bebe21b1ac0091bf9f4e2a3e788564c0a95b866dc624fbc1b49dab3) attribute returned by [`cuDeviceGetAttribute`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__DEVICE.html#group__CUDA__DEVICE_1g9c3e1414f0ad901d3278a4d6645fc266).

Reply:
Thank you! I think adding a new entry to the enum would be nice as it allows us to handle it accordingly in `detail::enqueue_kernel_launch::handleError` (if that is the way it goes, please add handling in here.)

@smaslov-intel - What are your thoughts on this? Should we add a new error code for this purpose? It could use `PI_PLUGIN_SPECIFIC_ERROR` but I am not sure how it would correctly report the cause with this.

Reply:
I think we should use `PI_PLUGIN_SPECIFIC_ERROR` as much as possible, unless a very generic error code is needed (and then add it to pi.h). With the `PI_PLUGIN_SPECIFIC_ERROR` you'd get error message with `piPluginGetLastError`: https://github.com/intel/llvm/blob/2e3f61e64e7d9147c40b5b75b0677124ec5ac781/sycl/include/CL/sycl/detail/pi.h#L1810 and return a non PI_SUCCESS from it to SYCL RT if any special action is expected there (i.e. if it's not a mere warning to the user).

Reply:
Have changed to use `PI_SPECIFIC_ERROR` but currently `piPluginGetLastError`(even when setting `SYCL_RT_WARNING_LEVEL=2`)  is not being called by the PI so investigating why this is.

Reply:
It seems that you are right. I believe this would be fixed by changing the default case in `detail::enqueue_kernel_launch::handleError` to call `checkPiResult` on the error code.

Reply:
Also, if you make that change, could you please change the docs description to specify the corresponding exception type?

Reply:
Yes that does it thanks @steffenlarsen ! Do we think that it is OK that the error message is only printed if `SYCL_RT_WARNING_LEVEL` is set?

Reply:
I think that should be fine. @smaslov-intel, what do you think?